### PR TITLE
Add pipeline(s) to rollback migrations for edxapp

### PR DESCRIPTION
**Pull request by macdiesel**
I've done a dry-run of this against the tools gocd and the xml config looks correct. I can post the non secret parts if we think that would help with the review.

### CODE REVIEW

**Comment by macdiesel** [comment]
FYI I'm currently working on cleaning up the docs on this.

File: edxpipelines/patterns/edxapp.py [hunk 1]
```diff
@@ -489,6 +483,127 @@ def rollback_asgs(edxapp_deploy_group, pipeline_name, build_pipeline, deploy_pip
...
+        asgard_api_endpoints
+        asgard_token
+        aws_access_key_id
+        aws_secret_access_key
+        hipchat_token
+    """
+    # In order to roll back the migrations:
+    # - Launch instance
+    # - LMS default/CSMH Rollback migrations
+    # - CMS default/CSMH Rollback migarations
```
Comment (doctoryes) [inline edxpipelines/patterns/edxapp.py]: Typo

File: edxpipelines/patterns/edxapp.py [hunk 2]
```diff
@@ -489,6 +483,127 @@ def rollback_asgs(edxapp_deploy_group, pipeline_name, build_pipeline, deploy_pip
...
+            application_name=config['play_name'],
+            application_path=config['application_path'],
+            sub_application_name=sub_app
+        )
+
+    launch_info_location = utils.ArtifactLocation(
+        pipeline.name,
+        constants.LAUNCH_INSTANCE_STAGE_NAME,
+        constants.LAUNCH_INSTANCE_JOB_NAME,
+        constants.LAUNCH_INSTANCE_FILENAME
```
Comment (doctoryes) [inline edxpipelines/patterns/edxapp.py]: General comment: It'd be clearer that the artifact is being fetched from the earlier launch_instance stage if it got these values from `launch_stage`.

File: edxpipelines/patterns/jobs.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        "pipeline": instance_key_location.pipeline,
+        "stage": instance_key_location.stage,
+        "job": instance_key_location.job,
+        "src": FetchArtifactFile(instance_key_location.file_name),
+        "dest": constants.ARTIFACT_PATH
+    }
+    job.add_task(FetchArtifactTask(**artifact_params))
+
+    # ensure the target directoy exists
+    tasks.generate_target_directory(job)
```
Comment (doctoryes) [inline edxpipelines/patterns/jobs.py]: Nit: Move this before or after the three artifact fetches.

**Comment by macdiesel** [comment]
Comments addressed.

**Comment by doctoryes** [comment]
👍

File: edxpipelines/patterns/edxapp.py [hunk 4]
```diff
@@ -118,12 +118,6 @@ def build_migrate_deploy_subset_pipeline(
             config['edx_environment'],
             base_ami_id
         )
-        ami_artifact = utils.ArtifactLocation(
-            pipeline.name,
-            constants.BASE_AMI_SELECTION_STAGE_NAME,
-            constants.BASE_AMI_SELECTION_JOB_NAME,
-            FetchArtifactFile(constants.BASE_AMI_OVERRIDE_FILENAME),
-        )
```
Comment (cpennington) [inline edxpipelines/patterns/edxapp.py]: Why did you delete this?

File: edxpipelines/patterns/edxapp.py [hunk 5]
```diff
@@ -489,6 +483,134 @@ def rollback_asgs(edxapp_deploy_group, pipeline_name, build_pipeline, deploy_pip
...
+            stage_name=constants.BASE_AMI_SELECTION_STAGE_NAME,
+            material_name='select_base_ami',
+        )
+    )
+
+    # Create the armed stage as this pipeline needs to auto-execute
+    stages.generate_armed_stage(pipeline, constants.ARMED_JOB_NAME)
+
+    # launch instance
+    launch_stage = stages.generate_launch_instance(
```
Comment (cpennington) [inline edxpipelines/patterns/edxapp.py]: The `build_migrate_deploy_subset_pipeline` function already packages up the launch-execute-cleanup pattern. Would it make sense to just call that?

File: edxpipelines/patterns/jobs.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+    # Fetch the SSH key to use in reaching the EC2 instance.
+    artifact_params = {
+        "pipeline": instance_key_location.pipeline,
+        "stage": instance_key_location.stage,
+        "job": instance_key_location.job,
+        "src": FetchArtifactFile(instance_key_location.file_name),
+        "dest": constants.ARTIFACT_PATH
+    }
+    job.add_task(FetchArtifactTask(**artifact_params))
```
Comment (cpennington) [inline edxpipelines/patterns/jobs.py]: This seems like a pattern that gets re-used a lot. Does it make sense to pull it out as a separate function that just sets up the relevant keys?

File: edxpipelines/patterns/stages.py [hunk 7]
```diff
@@ -710,6 +713,69 @@ def generate_run_migrations(pipeline,
...
+            'DB_MIGRATION_USER': 'migrate',
+            'ARTIFACT_PATH': constants.ARTIFACT_PATH,
+            'ANSIBLE_CONFIG': constants.ANSIBLE_CONTINUOUS_DELIVERY_CONFIG,
+        }
+    )
+    pipeline.ensure_encrypted_environment_variables(
+        {
+            'DB_MIGRATION_PASS': db_migration_pass,
+        }
+    )
```
Comment (cpennington) [inline edxpipelines/patterns/stages.py]: Can you push these down to the tasks that actually use them?

File: edxpipelines/patterns/tasks.py [hunk 8]
```diff
@@ -301,6 +301,79 @@ def generate_check_migration_duration(job,
...
+    generate_target_directory(job, migration_artifact_path)
+    job.ensure_artifacts(
+        set(
+            [BuildArtifact(migration_artifact_path)]
+        )
+    )
+
+    command = ' '.join(
+        [
+            'for migration_input_file in ../{migration_artifact_path}/*_migration_plan.yml do',
```
Comment (cpennington) [inline edxpipelines/patterns/tasks.py]: Rather than looping in bash, would it make sense to make multiple tasks, each of which runs one migration plan?

Comment (macdiesel) [inline edxpipelines/patterns/tasks.py]: I mulled over a few options but ultimately pushing this down to bash allowed us to not need to know the database/file names in the gomatic code and allows this to cover all databases for any IDA without prior knowledge of what exists for that application.

File: edxpipelines/patterns/stages.py (hunk 7 above)
Comment (macdiesel) [inline edxpipelines/patterns/stages.py]: I could but we currently do not pass the pipelines down to the tasks/jobs. If we are going to refactor this later I would rather hold this work until that point. 

Also Job does not offer ensure_unencrypted_environment_variable. Would like to know a bit more about this before we start pushing these from the pipeline level to the job level.

File: edxpipelines/patterns/jobs.py (hunk 6 above)
Comment (macdiesel) [inline edxpipelines/patterns/jobs.py]: Willing to do this as a follow on to this PR.

File: edxpipelines/patterns/edxapp.py (hunk 5 above)
Comment (macdiesel) [inline edxpipelines/patterns/edxapp.py]: I will take a look.

File: edxpipelines/patterns/jobs.py (hunk 6 above)
Comment (cpennington) [inline edxpipelines/patterns/jobs.py]: Cool.

File: edxpipelines/patterns/stages.py (hunk 7 above)
Comment (cpennington) [inline edxpipelines/patterns/stages.py]: Fair. In another PR that I started working on, I've been passing both pipelines and jobs down into tasks. But we can postpone this refactor until that work happens.

File: edxpipelines/patterns/tasks.py (hunk 8 above)
Comment (cpennington) [inline edxpipelines/patterns/tasks.py]: Ah, got it.

File: edxpipelines/patterns/edxapp.py (hunk 4 above)
Comment (macdiesel) [inline edxpipelines/patterns/edxapp.py]: user error, adding back in.

File: edxpipelines/patterns/edxapp.py (hunk 5 above)
Comment (macdiesel) [inline edxpipelines/patterns/edxapp.py]: @cpennington I have refactored this piece of code to use the subset function, could you do a quick re-review?

File: edxpipelines/patterns/jobs.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+    # fetch the migration outputs
+    artifact_params = {
+        "pipeline": migration_info_location.pipeline,
+        "stage": migration_info_location.stage,
+        "job": migration_info_location.job,
+        "src": FetchArtifactDir(migration_info_location.file_name),
+        "dest": constants.ARTIFACT_PATH
+    }
+    job.add_task(FetchArtifactTask(**artifact_params))
```
Comment (cpennington) [inline edxpipelines/patterns/jobs.py]: Nit: You can use `migration_info_location.as_fetch_task(constants.ARTIFACT_PATH)` for this now (and construct `migration_info_location` with `is_dir` set to `True`).

Comment (macdiesel) [inline edxpipelines/patterns/jobs.py]: good stuff.
